examples/palace-tonegen.py

```diff
@@ -0,0 +1,127 @@
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with palace.  If not, see <https://www.gnu.org/licenses/>.
+
+from argparse import Action, ArgumentParser
+from array import array
+from math import pi, sin
+from random import random
+from typing import Iterable, Tuple
+
+from palace import Buffer, Context, BaseDecoder, Device
+
+from numpy import linspace
+from numpy.random import random_sample
+from scipy import signal
+
+PERIOD: float = 0.025
+WAVEFORMS = {'sine': sine,
+             'square': square,
+             'sawtooth': sawtooth,
+             'triangle': triangle,
+             'impulse': impulse,
+             'white noise': white_noise}
+
+
+class ToneGenerator(BaseDecoder):
+    def __init__(self, waveform: str):
+        self.func = lambda frame: WAVEFORMS[waveform](frame/self.frequency,
+                                           self.frequency)
+
+    @BaseDecoder.frequency.getter
+    def frequency(self) -> int: return 44100
+
+    @BaseDecoder.channel_config.getter
+    def channel_config(self) -> str:
+        return 'Mono'
+
+    @BaseDecoder.sample_type.getter
+    def sample_type(self) -> str:
+        return 'Float32'
+
+    @BaseDecoder.length.getter
+    def length(self) -> int: return 0
+
+    def seek(self, pos: int) -> bool: return False
+
+    @BaseDecoder.loop_points.getter
+    def loop_points(self) -> Tuple[int, int]: return 0, 0
+
+    def read(self, count: int) -> bytes:
+        stop = min(self.start + count, self.length)
+        data = array('f', map(self.func, range(self.start, stop)))
+        self.start = stop
+        return data.tobytes()
+
+
+class TypePrinter(Action):
+    def __call__(self, parser: ArgumentParser, *args, **kwargs) -> None:
+        print('Available waveform types:', *WAVEFORMS, sep='\n')
+        parser.exit()
+
+
+def sine(time: int, sample_rate: int) -> Iterable[float]:
+    """Generate sinusoidal signal."""
+    t = linspace(0, time, time * sample_rate, endpoint=False)
+    data = sin(t)
+    return data
+
+
+def square(time: int, sample_rate: int) -> Iterable[float]:
+    """Generate square signal."""
+    t = linspace(0, time, time * sample_rate, endpoint=False)
+    data = signal.square(2 * pi * 5 * t)
+
+
+def sawtooth(time: int, sample_rate: int) -> Iterable[float]:
+    t = linspace(0, time, time * sample_rate, endpoint=False)
+    data = signal.sawtooth(2 * pi * 5 * t)
+
+
+def triangle(time: int, sample_rate: int) -> Iterable[float]:
+    pass
+
+
+def impulse(time: int, sample_rate: int) -> Iterable[float]:
+    pass
+
+
+def white_noise(time: int, sample_rate: int) -> Iterable[float]:
+    return random_sample((1, time * sample_rate))
+    
+
+
+def play(device: str, waveform: str, duration: int) -> None:
+    with Device(device) as dev, Context(dev):
+        dec = ToneGenerator(waveform)
+        _ = duration  # use it later
```

Comment:
What do you thing that this will really do :smile:?  BTW for the tone generator to be defined, there needs to be 3 parameters:
1. waveform
2. frequency, i.e. pitch
3. duration, since we want it to be finite